torch/jit/frontend.py

```diff
@@ -221,6 +221,18 @@ def remove_prefix(text, prefix):
     aligned_prefix.append(fn_def)
     return aligned_prefix + aligned_suffix
 
+def _replace_events_and_streams(sourcelines):
+    """
+    This helper function searches for ``torch.cuda.Stream`` and ``torch.cuda.Event``
+    and replaces it with the custom classes ``torch.classes.cuda.Stream`` and
+    ``torch.classes.cuda.Event`` respectively. This is done because CUDA streams and events
+    are implemented as custom classes in JIT.
+    """
+    for i in range(len(sourcelines)):
+        sourcelines[i] = sourcelines[i].replace("torch.cuda.Stream", "torch.classes.cuda.Stream")
+        sourcelines[i] = sourcelines[i].replace("torch.cuda.Event", "torch.classes.cuda.Event")
+
+    return sourcelines
```

Comment:
Is it possible to add this logic in the type resolver/related type resolution code? Something similar to how we do type resolution for `ScriptClass`:

https://github.com/pytorch/pytorch/blob/65087dd1d4c1e503281cd676e5563e9e64b2a926/torch/csrc/jit/python/python_sugared_value.cpp#L1161-L1165

Reply:
I haven't tried this. Let me try it out.

Reply:
`torch.cuda.Stream` calls the eager mode `Stream` if we do not replace it in `frontend.py`. So this won't be possible to defer the type resolution later in  `python_sugared_value.cpp`.

Reply:
Hmm, but I think that happens because `CUDAModuleValue::attr` uses `PythonValue::attr` when the attribute is not one of the functions we have a custom operator for, right? 

https://github.com/pytorch/pytorch/blob/b936abd840d8fa3122d3a6166c0d7344d8352901/torch/csrc/jit/python/python_sugared_value.cpp#L235-L253

Reply:
Yeah, so that gets resolved as object of class torch.cuda.streams.Stream 

`obj name::<class 'torch.cuda.streams.Stream'>
I enabled a few debug prints , so a call to `torch.cuda.Stream()` results in these values:
 
```
obj name::<module 'torch' from '/data/users/nikithamalgi/pytorch/torch/__init__.py'>
name::::torch
obj name::<module 'torch.cuda' from '/data/users/nikithamalgi/pytorch/torch/cuda/__init__.py'>
name::::torch.cuda
field:::Stream
**obj name::<class 'torch.cuda.streams.Stream'>**
```
Here `torch.cuda` results in `CUDAModuleValue::attr` while `torch.classes.cuda` results in `PythonValue::getattr`.

Reply:
So we should be able to return an instance of `ClassValue` corresponding to the custom class that we registered for `Stream` when TorchScript tries to use `torch.cuda.Stream`, right?

This doesn't handle resolving type annotations, but it's part of the way there, right?

Reply:
I tried returning `ClassValue` corresponding to custom class and I ended up in compilation errors.

Here is what I tried:
```
return std::make_shared<PythonClassValue>(
        py::cast<torch::jit::CUDAStream>(member));
./torch/csrc/jit/python/python_sugared_value.h:300:26: note:   no known conversion for argument 1 from ‘torch::jit::CUDAStream’ to ‘const torch::jit::PythonClassValue&’
```
When `torch.classes.cuda.Stream` is used, it gets resolved as `torch._C.ScriptClass`.

Reply:
Oh, but that's a `PythonClassValue`, and we don't have a Python object in this case. What about a regular `ClassValue`?

Reply:
Yes regular `ClassValue` also resulted in the same error :
```
../torch/csrc/jit/frontend/sugared_value.h:335:18: note:   no known conversion for argument 1 from ‘torch::jit::CUDAStream’ to ‘const torch::jit::ClassValue&’
```

Reply:
The constructor for `ClassValue` accepts a `ClassType`, not a `jit::CUDAStream`.

Reply:
More specifically, it is a `ClassTypePtr` (i.e. `std::shared_ptr<ClassType>`), not a raw `ClassType` instance.

Reply:
OK `ClassTypePtr` compiled, but I ran into RuntimeError:

```
RuntimeError: Unable to cast Python instance of type <class 'type'> to C++ type 'std::shared_ptr<c10::ClassType>'
```

Reply:
It seems like you tried to cast the Python object representing `torch.cuda.Stream` into a `ClassTypePtr`? You should look up the `ClassType` for `jit::Stream` in the compilation unit and use that.

Reply:
As suggested, tried the above. But I'm still running into errors.
Here is how I did it:
```
    auto pyCu = get_python_cu();
    auto qualname = c10::QualifiedName("torch::jit::CUDAStream");
    auto type = pyCu->get_type(qualname);
    return std::make_shared<ClassValue>(type);
```

Also, `torch.classes.cuda.Stream` resolves to  `torch._C.ScriptClass`. I also tried replacing `torch::jit::CUDAStream` with `torch.classes.cuda.Stream`, that did not work either.

Reply:
Isn't the fully qualified name `__torch__.torch.classes.cuda.Stream`?

Reply:
Yes, `__torch__.torch.classes.cuda.Stream` that did not work either. I tried making it a PythonClassValue as well, didn't work.

Reply:
Can you dump all of the registered classes in the compilation unit? Maybe we're missing something.

When you try `__torch__.torch.classes.cuda.Stream`, is the class type successfully found? Is it the same error as before?

Reply:
Registeration is the issue here. `__torch__.torch.classes.cuda.Stream` was not registered.  I was able to resolve this, but I am still running into errors like: 
```
_jit_script_compile qualname --__torch__.fn
find_function:__torch__.fn
field:::Stream
Registered classes for :::__torch__.torch.classes.cuda.Stream
{__torch__.torch.classes.cuda.Stream: 0}
it->second0x559dee724130
Traceback (most recent call last):
  File "repro.py", line 4, in <module>
    def fn():
  File "/data/users/nikithamalgi/pytorch/torch/jit/_script.py", line 994, in script
    fn = torch._C._jit_script_compile(
RuntimeError: Unable to cast Python instance of type <class 'type'> to C++ type 'torch::jit::ScriptClass'
```
I'm not sure why class results to `type` here. 

The code is:
```
    auto pyCu = get_python_cu();
    auto qualname = c10::QualifiedName("__torch__.torch.classes.cuda.Stream");
    auto class_type = ClassType::create(qualname, pyCu);
    pyCu->register_type(class_type);
    return std::make_shared<ClassValue>(class_type);
```

Reply:
The `ClassType` should already exist, so you should not create it like that. 

Can you add some debug code to `CompilationUnit` to print out all the types so that we can determine if the type name is being spelling correctly? TorchBind uses static registration, so the type should already be registered.

Reply:
Well the register class types return empty.

Here is a log

```
_jit_script_compile qualname --__torch__.fn
find_function:__torch__.fn
All registered class types   # This did not print anything
Register Function - __torch__.fn
field:::Stream
```

Reply:
> TorchBind uses static registration, so the type should already be registered.

I was right, but not about *where* the types are stored. Can you try using `getCustomClass()` instead of `CompiltationUnit::get_type`.

Reply:
Fixed now.